mmap-queue/src/main/java/org/tools4j/mmap/queue/impl/DefaultAppenderIdPool.java

```diff
@@ -0,0 +1,147 @@
+
+/**
+ * A pool of appender ids based on a memory-mapped file associated with a given queue name.
+ * It guarantees that there are no two or more processes/threads acquiring the same id for given queue name.
+ *
+ * File contains number of open appenders (counter). Acquiring of an appender id is done
+ * by incrementing the counter atomically and releasing appender id is done
+ * by decrementing the counter atomically.
+ */
+public class DefaultAppenderIdPool implements AppenderIdPool {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAppenderIdPool.class);
+
+    private static final int MAX_APPENDERS = 256;
+    private static final int REGION_SIZE = 8;
+    private static final String FILE_SUFFIX = "open_appenders";
+    private final RegionAccessor region;
+    private final UnsafeBuffer buffer = new UnsafeBuffer();
+    private final String queueName;
+    private volatile boolean closed = false;
+
+    public DefaultAppenderIdPool(final String directory, final String queueName) {
+        this.queueName = requireNonNull(queueName);
+        region = FixedSizeRegion.open(directory, queueName + "_" + FILE_SUFFIX,
+                REGION_SIZE, fileInitialiser());
+
+        region.wrap(0, buffer);
+    }
+
+    @Override
+    public short acquire() {
+        if (closed) {
+            throw new IllegalStateException(format("Appender id pool for %s queue is closed", queueName));
+        }
+
+        int currentAppenderCounter;
+        do {
+            currentAppenderCounter = buffer.getIntVolatile(0);
+            if (currentAppenderCounter >= MAX_APPENDERS) {
+                throw new IllegalStateException(format("Exceeded max number of appenders %d in %s queue", MAX_APPENDERS, queueName));
+            }
+        } while (!buffer.compareAndSetInt(0, currentAppenderCounter, currentAppenderCounter + 1));
+
+        LOGGER.info("Acquired appenderId {} for {} queue", currentAppenderCounter, queueName);
+        return (short) currentAppenderCounter;
+    }
+
+    @Override
+    public void release(final short appenderId) {
+        if (closed) {
+            throw new IllegalStateException(format("Appender id pool for %s queue is closed", queueName));
+        }
+
+        int currentAppenderCounter;
+        do {
+            currentAppenderCounter = buffer.getIntVolatile(0);
+        } while (!buffer.compareAndSetInt(0, currentAppenderCounter, currentAppenderCounter - 1));
+        LOGGER.info("Released appenderId {} for {} queue", appenderId, queueName);
+    }
+
+    @Override
+    public void close() {
+        this.closed = true;
+        int currentCounter = buffer.getIntVolatile(0);
+        region.close();
+        LOGGER.info("Closed appender id pool. queue={}, open appenders={}", queueName, currentCounter);
+    }
+
+    private static FileInitialiser fileInitialiser() {
+        return (fileName, fileChannel) -> {
+            try {
+                FileLock fileLock = acquireLock(fileChannel);
```

Comment:
could potentially block forever -- maybe specify max attempts to acquire lock in a constant and throw exception